File: tests/helpers/fixture.ts
import { promises as fs } from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

/**
 * Spec for an in-tree fixture. Profiles live under `.claude-profiles/<name>/`
 * with `profile.json` and a `.claude/` subtree of arbitrary file content.
 */
export interface ProfileSpec {
  /** profile.json content (object will be stringified). If a string, written verbatim. */
  manifest?: Record<string, unknown> | string | null;
  /** Files under `.claude/`, keyed by relative posix path → content string. */
  files?: Record<string, string>;
  /**
   * Files at the profile root (peer of profile.json, sibling of .claude/),
   * keyed by relative posix path → content string. Used for cw6 to set up
   * `.claude-profiles/<P>/CLAUDE.md` (the destination='projectRoot' source).
   */
  rootFiles?: Record<string, string>;
}

export interface ComponentSpec {
  files?: Record<string, string>;
  /**
   * Files at the component root (sibling of .claude/). For cw6 testing of
   * include contributors that supply a profile-root CLAUDE.md.
   */
  rootFiles?: Record<string, string>;
}

export interface FixtureSpec {
  profiles?: Record<string, ProfileSpec>;
  /** In-repo components (under `.claude-profiles/_components/<name>/`). */
  components?: Record<string, ComponentSpec>;
  /** Out-of-repo paths (absolute keys). */
  external?: Record<string, ComponentSpec>;
}

export interface Fixture {
  projectRoot: string;
  externalRoot: string;
  cleanup: () => Promise<void>;
}

export async function makeFixture(spec: FixtureSpec): Promise<Fixture> {
  const tmp = await fs.mkdtemp(path.join(os.tmpdir(), "ccp-fixture-"));
  const projectRoot = path.join(tmp, "project");
  const externalRoot = path.join(tmp, "external");
  await fs.mkdir(projectRoot, { recursive: true });
  await fs.mkdir(externalRoot, { recursive: true });

  const profilesDir = path.join(projectRoot, ".claude-profiles");

  for (const [name, p] of Object.entries(spec.profiles ?? {})) {
    const dir = path.join(profilesDir, name);
    await fs.mkdir(dir, { recursive: true });
    if (p.manifest !== null && p.manifest !== undefined) {
      const content =
        typeof p.manifest === "string" ? p.manifest : JSON.stringify(p.manifest, null, 2);
      await fs.writeFile(path.join(dir, "profile.json"), content);
    }
    for (const [rel, content] of Object.entries(p.files ?? {})) {
      const fp = path.join(dir, ".claude", rel);
      await fs.mkdir(path.dirname(fp), { recursive: true });
      await fs.writeFile(fp, content);
    }
    for (const [rel, content] of Object.entries(p.rootFiles ?? {})) {
      const fp = path.join(dir, rel);
      await fs.mkdir(path.dirname(fp), { recursive: true });
      await fs.writeFile(fp, content);
    }
  }

  for (const [name, c] of Object.entries(spec.components ?? {})) {
    const dir = path.join(profilesDir, "_components", name);
    await fs.mkdir(dir, { recursive: true });
    for (const [rel, content] of Object.entries(c.files ?? {})) {
      const fp = path.join(dir, ".claude", rel);
      await fs.mkdir(path.dirname(fp), { recursive: true });
      await fs.writeFile(fp, content);
    }
    for (const [rel, content] of Object.entries(c.rootFiles ?? {})) {
      const fp = path.join(dir, rel);
      await fs.mkdir(path.dirname(fp), { recursive: true });
      await fs.writeFile(fp, content);
    }
  }

  for (const [absKey, c] of Object.entries(spec.external ?? {})) {
    // The "absolute key" is treated as a path under externalRoot to keep
    // tests hermetic. Tests pass `path.join(externalRoot, absKey)` to the
    // resolver as the includes entry to point at this directory.
    const dir = path.join(externalRoot, absKey);
    await fs.mkdir(dir, { recursive: true });
    for (const [rel, content] of Object.entries(c.files ?? {})) {
      const fp = path.join(dir, ".claude", rel);
      await fs.mkdir(path.dirname(fp), { recursive: true });
      await fs.writeFile(fp, content);
    }
  }

  return {
    projectRoot,
    externalRoot,
    cleanup: () => fs.rm(tmp, { recursive: true, force: true }),
  };
}

File: tests/integration/helpers/fixture.go
// Package helpers provides the Go integration-test harness: MakeFixture
// builds an isolated project root under t.TempDir() and RunCli spawns the
// built c3p binary against it.
//
// Every Fixture uses t.TempDir() for isolation; tests never share state.
package helpers

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"testing"
)

// ProfileSpec describes an in-tree profile fixture. Profiles live under
// .claude-profiles/<name>/ with profile.json and a .claude/ subtree of
// arbitrary file content.
type ProfileSpec struct {
	// Manifest may be a map (will be JSON-marshalled), a string (written
	// verbatim — the test wants invalid JSON), or nil (no profile.json
	// written — the test wants the manifest absent).
	Manifest any
	// Files keys relative posix paths under .claude/ → file content.
	Files map[string]string
	// RootFiles keys relative posix paths at the profile root (sibling of
	// .claude/) → file content. Used for cw6 to set up
	// .claude-profiles/<P>/CLAUDE.md (the destination='projectRoot' source).
	RootFiles map[string]string
}

// ComponentSpec describes an in-repo or out-of-repo component fixture.
type ComponentSpec struct {
	Files     map[string]string
	RootFiles map[string]string
}

// FixtureSpec is the input to MakeFixture: profiles, components, and
// out-of-repo paths that tests treat as absolute.
type FixtureSpec struct {
	Profiles map[string]ProfileSpec
	// Components live under .claude-profiles/_components/<name>/.
	Components map[string]ComponentSpec
	// External keys relative paths under ExternalRoot (the test treats
	// them as "absolute" because the resolver sees a fully-qualified path).
	External map[string]ComponentSpec
}

// Fixture is the handle returned by MakeFixture. ProjectRoot is the
// in-repo root the CLI runs against (the --cwd target); ExternalRoot is
// the parallel tree for out-of-repo includes. The temp tree is cleaned up
// automatically by t.Cleanup, including on test panic.
type Fixture struct {
	ProjectRoot  string
	ExternalRoot string
}

// Cleanup mirrors the TS fixture.cleanup surface to keep the F1 explicit
// contract (Fixture{ProjectRoot, ExternalRoot, Cleanup() error}) honest
// across both languages — the helper-parity audit (PR4) checks for this
// method by name. The Go implementation is intentionally a no-op: every
// MakeFixture caller owns a t.TempDir() root, and Go's testing runtime
// removes that tree at test end (including on panic), so re-running rm
// here would be redundant. Returns nil for symmetry with the TS Promise
// surface; never returns an error.
func (*Fixture) Cleanup() error { return nil }

// MakeFixture materializes spec into a fresh t.TempDir() tree and returns
// a Fixture. The tree is removed at test end automatically (including on
// panic), since t.TempDir() registers its own cleanup. Errors during
// creation surface via t.Fatal — matches the TS pattern of "throw and the
// test bails".
func MakeFixture(t *testing.T, spec FixtureSpec) *Fixture {
	t.Helper()
	f, err := makeFixtureAt(t.TempDir(), spec)
	if err != nil {
		t.Fatalf("makeFixture: %v", err)
	}
	return f
}

func makeFixtureAt(tmp string, spec FixtureSpec) (*Fixture, error) {
	projectRoot := filepath.Join(tmp, "project")
	externalRoot := filepath.Join(tmp, "external")
	if err := os.MkdirAll(projectRoot, 0o755); err != nil {
		return nil, fmt.Errorf("mkdir project: %w", err)
	}
	if err := os.MkdirAll(externalRoot, 0o755); err != nil {
		return nil, fmt.Errorf("mkdir external: %w", err)
	}

	profilesDir := filepath.Join(projectRoot, ".claude-profiles")

	for name, p := range spec.Profiles {
		dir := filepath.Join(profilesDir, name)
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return nil, fmt.Errorf("mkdir profile %q: %w", name, err)
		}
		if err := writeManifest(filepath.Join(dir, "profile.json"), p.Manifest); err != nil {
			return nil, fmt.Errorf("profile %q manifest: %w", name, err)
		}
		if err := writeTree(filepath.Join(dir, ".claude"), p.Files); err != nil {
			return nil, fmt.Errorf("profile %q files: %w", name, err)
		}
		if err := writeTree(dir, p.RootFiles); err != nil {
			return nil, fmt.Errorf("profile %q rootFiles: %w", name, err)
		}
	}

	for name, c := range spec.Components {
		dir := filepath.Join(profilesDir, "_components", name)
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return nil, fmt.Errorf("mkdir component %q: %w", name, err)
		}
		if err := writeTree(filepath.Join(dir, ".claude"), c.Files); err != nil {
			return nil, fmt.Errorf("component %q files: %w", name, err)
		}
		if err := writeTree(dir, c.RootFiles); err != nil {
			return nil, fmt.Errorf("component %q rootFiles: %w", name, err)
		}
	}

	for absKey, c := range spec.External {
		dir := filepath.Join(externalRoot, absKey)
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return nil, fmt.Errorf("mkdir external %q: %w", absKey, err)
		}
		if err := writeTree(filepath.Join(dir, ".claude"), c.Files); err != nil {
			return nil, fmt.Errorf("external %q files: %w", absKey, err)
		}
		if err := writeTree(dir, c.RootFiles); err != nil {
			return nil, fmt.Errorf("external %q rootFiles: %w", absKey, err)
		}
	}

	return &Fixture{
		ProjectRoot:  projectRoot,
		ExternalRoot: externalRoot,
	}, nil
}

// writeManifest mirrors the TS three-way switch: nil → write nothing;
// string → write verbatim (test wants invalid/exotic JSON); any other →
// JSON-marshal with two-space indent (matches TS JSON.stringify(_, null, 2)
// byte-for-byte for object literals).
func writeManifest(path string, manifest any) error {
	if manifest == nil {
		return nil
	}
	switch v := manifest.(type) {
	case string:
		return os.WriteFile(path, []byte(v), 0o644)
	default:
		bytes, err := json.MarshalIndent(v, "", "  ")
		if err != nil {
			return err
		}
		return os.WriteFile(path, bytes, 0o644)
	}
}

func writeTree(root string, files map[string]string) error {
	for rel, content := range files {
		fp := filepath.Join(root, filepath.FromSlash(rel))
		if err := os.MkdirAll(filepath.Dir(fp), 0o755); err != nil {
			return err
		}
		if err := os.WriteFile(fp, []byte(content), 0o644); err != nil {
			return err
		}
	}
	return nil
}
